// datatables.net-vue3/dist/datatables.net-vue3.mjs
import { defineComponent as g, ref as d, watch as k, onMounted as x, getCurrentInstance as D, unref as _, onBeforeUnmount as j, setBlockTracking as v, createElementVNode as b, mergeProps as A, renderSlot as S } from "vue";
const y = [
  "childRow",
  "column-sizing",
  "column-visibility",
  "destroy",
  "draw",
  "error",
  "init",
  "length",
  "order",
  "page",
  "preDraw",
  "preInit",
  "preXhr",
  "processing",
  "requestChild",
  "search",
  "stateLoadParams",
  "stateLoaded",
  "stateSaveParams",
  "xhr",
  "autoFill",
  "preAutoFill",
  "buttons-action",
  "buttons-processing",
  "column-reorder",
  "key",
  "key-blur",
  "key-focus",
  "key-refocus",
  "key-return-submit",
  "responsive-display",
  "responsive-resize",
  "rowgroup-datasrc",
  "pre-row-reorder",
  "row-reorder",
  "row-reordered",
  "dtsb-inserted",
  "deselect",
  "select",
  "select-blur",
  "selectItems",
  "selectStyle",
  "user-select",
  "stateRestore-change"
];
let c;
const C = {
  name: "Datatables.netVue",
  inheritAttrs: !1,
  use(s) {
    c = s;
  }
}, E = /* @__PURE__ */ g({
  ...C,
  props: {
    ajax: null,
    columns: null,
    data: null,
    options: null
  },
  emits: y,
  setup(s, { expose: u }) {
    const r = s, f = d(null), a = d(), h = d([]);
    k(
      () => r.data,
      (e) => {
        var l, o, i, m;
        let t = (l = a.value) == null ? void 0 : l.data().toArray();
        for (let n of e)
          t != null && t.includes(n) || (o = a.value) == null || o.row.add(n);
        if (typeof t < "u")
          for (let n of t)
            e.includes(n) || (i = a.value) == null || i.row((I, w) => w === n).remove();
        (m = a.value) == null || m.rows().invalidate().draw(!1), p(e);
      },
      {
        deep: !0
      }
    ), x(() => {
      const e = D();
      let t = r.options || {};
      if (r.data && (t.data = r.data, p(t.data)), r.columns && (t.columns = r.columns), r.ajax && (t.ajax = r.ajax), !c)
        throw new Error(
          "DataTables library not set. See https://datatables.net/tn/19 for details."
        );
      a.value = new c(_(f), t);
      for (let l of y)
        a.value && e && a.value.on(l, function() {
          var o = Array.from(arguments), i = o.shift();
          o.unshift({ event: i, dt: a }), o.unshift(l), e.emit.apply(e, o);
        });
    }), j(() => {
      var e;
      (e = a.value) == null || e.destroy(!0);
    });
    function p(e) {
      h.value = e.value ? e.value.slice() : e.slice();
    }
    return u({
      dt: a
    }), (e, t) => t[0] || (v(-1), t[0] = b("div", { class: "datatable" }, [
      b("table", A({
        ref_key: "table",
        ref: f
      }, e.$attrs, { style: { width: "100%" } }), [
        S(e.$slots, "default")
      ], 16)
    ]), v(1), t[0]);
  }
}), P = /* @__PURE__ */ (() => {
  const s = E;
  return s.install = (u) => {
    u.component("Datatables.netVue", s);
  }, s;
})();
export {
  P as default
};
