// @ionic/cli/commands/live-update/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiveUpdatesNamespace = void 0;
const color_1 = require("../../lib/color");
const namespace_1 = require("../../lib/namespace");
class LiveUpdatesNamespace extends namespace_1.Namespace {
    async getMetadata() {
        return {
            name: 'live-update',
            summary: 'Ionic Live Updates functionality',
            description: `
These commands integrate with Ionic Cloud to configure the Live Updates plugin in your project and run remote builds.

Ionic Live Updates documentation:
- Overview: ${(0, color_1.strong)('https://ion.link/appflow-deploy-docs')}
`,
            groups: ["paid" /* MetadataGroup.PAID */],
        };
    }
    async getCommands() {
        return new namespace_1.CommandMap([
            ['add', async () => { const { AddCommand } = await Promise.resolve().then(() => require('./add')); return new AddCommand(this); }],
            ['configure', async () => { const { ConfigureCommand } = await Promise.resolve().then(() => require('./configure')); return new ConfigureCommand(this); }],
            ['manifest', async () => { const { LiveUpdatesManifestCommand } = await Promise.resolve().then(() => require('./manifest')); return new LiveUpdatesManifestCommand(this); }],
        ]);
    }
}
exports.LiveUpdatesNamespace = LiveUpdatesNamespace;
